[AArch64][Clang] Update predication of SVE2-AES/PMULL Pair Intrinsics and add Test Coverage

Amichaxx opened the pull request:
This patch updates the predication guards for SVE2 AES and PMULL Pair intrinsics to support execution in streaming mode.

- Updated tablegen (arm_sve.td)
- Updated existing Sema tests' diagnostics to reflect updated guard
- Updated runlines of acle_sve2_pmullb_128.c, acle_sve2_pmullt_128.c, sve2-intrinsics-polynomial-arithmetic-128.ll, acle_sve2_aesimc.c, acle_sve2_aesd.c, acle_sve2_aese.c and acle_sve2_aesmc.c to reflect streaming.

Amichaxx committed:
[AArch64][Clang] Update predication of SVE-AES/PMULL Pair Intrinsics and add test coverage

This patch updates the predication guards for SVE2 AES and PMULL Pair intrinsics to support execution in streaming mode.

- Updated tablegen (arm_sve.td)
- Updated existing llvm/clang codegen tests to support streaming
- Updated existing Sema tests' diagnostics to reflect updated guard

Lukacma commented on clang/test/CodeGen/AArch64/sve2-intrinsics/acle_sve2_aesimc.c:
```diff
@@ -1,8 +1,14 @@
-// RUN: %clang_cc1 -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve-aes -O1 -Werror -Wall -emit-llvm -o - -x c++ %s | FileCheck %s -check-prefix=CPP-CHECK
-// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve-aes -O1 -Werror -Wall -emit-llvm -o - %s | FileCheck %s
-// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve-aes -O1 -Werror -Wall -emit-llvm -o - -x c++ %s | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -fclang-abi-compat=latest -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve-aes -O1 -Werror -Wall -emit-llvm -o - %s | FileCheck %s
```
For future reference, it is better not to change existing runlines in tests if not necessary, as any change in flags will change the compiler behaviour and this purpose of the test. In this case it is fine though I think. Specifying abi is unrelated to what is tested here.

Lukacma replied:
Sorry, for not making myself clearer. You still need to add the new runlines for the ssve-aes flag. In my comments I was talking about existing runlines and runlines for sme-fa64 feature

Amichaxx replied:
Ah, I had interpreted it as any additional runlines being redundant. I've reapplied the relevant runlines. I've removed my changes to existing ones in acle_sve2_aesmic.c. They were initially changed to match the runlines of similar tests.

Lukacma commented on clang/test/CodeGen/AArch64/sve2-intrinsics/acle_sve2_pmullb_128.c:
```diff
@@ -6,6 +6,16 @@
+// RUN: %clang_cc1 -fclang-abi-compat=latest -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sme -target-feature +ssve-aes -O1 -Werror -Wall -emit-llvm -o - %s | FileCheck %s
+// RUN: %clang_cc1 -fclang-abi-compat=latest -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sme -target-feature +ssve-aes -O1 -Werror -Wall -emit-llvm -o - -x c++ %s | FileCheck %s -check-prefix=CPP-CHECK
+
+// RUN: %clang_cc1 -fclang-abi-compat=latest -triple aarch64 -target-feature +sme -target-feature +sme-fa64 -target-feature +sve-aes -O1 -Werror -Wall -emit-llvm -o - %s | FileCheck %s
```
These runlines are unnecessary.

Amichaxx replied:
Done.

Lukacma commented on clang/test/CodeGen/AArch64/sve2-intrinsics/acle_sve2_pmullt_128.c:
```diff
@@ -6,6 +6,16 @@
+// RUN: %clang_cc1 -fclang-abi-compat=latest -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sme -target-feature +ssve-aes -O1 -Werror -Wall -emit-llvm -o - %s | FileCheck %s
+// RUN: %clang_cc1 -fclang-abi-compat=latest -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sme -target-feature +ssve-aes -O1 -Werror -Wall -emit-llvm -o - -x c++ %s | FileCheck %s -check-prefix=CPP-CHECK
+
+// RUN: %clang_cc1 -fclang-abi-compat=latest -triple aarch64 -target-feature +sme -target-feature +sme-fa64 -target-feature +sve-aes -O1 -Werror -Wall -emit-llvm -o - %s | FileCheck %s
```
These runlines are unnecessary.

Lukacma commented on llvm/test/CodeGen/AArch64/sve2-intrinsics-polynomial-arithmetic-128.ll:
```diff
@@ -1,5 +1,7 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
 ; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sve2-aes < %s | FileCheck %s
+; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+ssve-aes -force-streaming < %s | FileCheck %s
+; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sve-aes,+sme-fa64 -force-streaming < %s | FileCheck %s
```
These runlines is unnecessary.

Amichaxx commented on clang/include/clang/Basic/arm_sve.td:
```diff
@@ -1828,17 +1828,19 @@ let SVETargetGuard = "sve2,lut", SMETargetGuard = "sme2,lut" in {
 
-def SVPMULLB_PAIR_U64   : SInst<"svpmullb_pair[_{d}]",   "ddd", "Ul", MergeNone, "aarch64_sve_pmullb_pair">;
-def SVPMULLB_PAIR_N_U64 : SInst<"svpmullb_pair[_n_{d}]", "dda", "Ul", MergeNone, "aarch64_sve_pmullb_pair">;
+let SVETargetGuard = "sve2,sve-aes", SMETargetGuard = "ssve-aes|sve-aes,sme-fa64" in {
```
This has been removed in the latest commit.

Amichaxx committed:
- Removed redundant run lines in acle_sve2_pmullb_128.c, sve2-intrinsics/acle_sve2_pmullt_128.c and sve2-intrinsics-polynomial-arithmetic-128.ll
- Removed redundant guard in arm_sve.td
- Reverted non-streaming runlines in acle_sve2_aesimc.c

Amichaxx committed:
- Removed redundant run lines in acle_sve2_pmullb_128.c, sve2-intrinsics/acle_sve2_pmullt_128.c and sve2-intrinsics-polynomial-arithmetic-128.ll
- Updated guard in arm_sve.td
- Reverted original run lines in acle_sve2_aesimc.c

Amichaxx commented on clang/include/clang/Basic/arm_sve.td:
```diff
@@ -1828,17 +1828,19 @@ let SVETargetGuard = "sve2,lut", SMETargetGuard = "sme2,lut" in {
 
-def SVPMULLB_PAIR_U64   : SInst<"svpmullb_pair[_{d}]",   "ddd", "Ul", MergeNone, "aarch64_sve_pmullb_pair">;
-def SVPMULLB_PAIR_N_U64 : SInst<"svpmullb_pair[_n_{d}]", "dda", "Ul", MergeNone, "aarch64_sve_pmullb_pair">;
+let SVETargetGuard = "sve2,sve-aes", SMETargetGuard = "ssve-aes" in {
```
They've been merged into the same block now.

Amichaxx commented:
Thank you for reviewing @Lukacma.

Lukacma commented:
I take it this can be merged now ?

Amichaxx commented:
I believe so. I don't have merge access, though.
